spec/index.bs

```diff
@@ -4168,6 +4173,113 @@ enum GPUStoreOp {
 };
 </script>
 
+## Rasterization state ## {#render-pass-encoder-rasterization-state}
+
+The render pass encoder has several methods which affect how draw commands are rasterized to the
+attachments used by this encoder.
+
+### <dfn method for=GPURenderPassEncoder>setViewport(x, y, width, height, minDepth, maxDepth)</dfn> ### {#GPURenderPassEncoder-setViewport}
+
+<div algorithm="GPURenderPassEncoder.setViewport">
+    <strong>|this|:</strong> of type {{GPURenderPassEncoder}}.
+
+    **Arguments:**
+        - `float` x - Left side of the viewport in pixels
+        - `float` y - Top side of the viewport in pixels
+        - `float` |width| of the viewport in pixels
+        - `float` |height| of the viewport in pixels
+        - `float` |minDepth| - Minimum depth value of the viewport
+        - `float` |maxDepth| - Maximum depth value of the viewport. May be less than or equal to
+            |minDepth|.
+
+    **Returns:** void
+
+    Sets the {{GPURenderPassEncoder/[[viewport]]}} used during the rasterization stage to linearly
+    map from normalized device coordinates to viewport coordinates.
+
+    On the [=Device timeline=], the following steps occur:
+
+        - If the [$GPURenderPassEncoder.setViewport/setViewport Valid Usage$] rules are met:
+
+            - Set |this|.{{GPURenderPassEncoder/[[viewport]]}} to the extents passed to the method.
+
+    <div class=validusage dfn-for=GPURenderPassEncoder.setViewport>
+        <dfn abstract-op>setViewport Valid Usage</dfn>
+
+        1. |width| must be greater than `0`.
+        1. |height| must be greater than `0`.
+        1. |minDepth| must be greater than or equal to `0.0` and less than or equal to `1.0`.
+        1. |maxDepth| must be greater than or equal to `0.0` and less than or equal to `1.0`.
+
+    </div>
+
+    Issue: Allowed for GPUs to use fixed point or rounded viewport coordinates
+
+</div>
+
+### <dfn method for=GPURenderPassEncoder>setScissorRect(x, y, width, height)</dfn> ### {#GPURenderPassEncoder-setScissorRect}
+
+<div algorithm="GPURenderPassEncoder.setScissorRect">
+    <strong>|this|:</strong> of type {{GPURenderPassEncoder}}.
+
+    **Arguments:**
+        - {{GPUIntegerCoordinate}} x - Left side of the scissor rectangle in pixels
+        - {{GPUIntegerCoordinate}} y - Top side of the scissor rectangle in pixels
+        - {{GPUIntegerCoordinate}} |width| of the scissor rectangle in pixels
+        - {{GPUIntegerCoordinate}} |height| of the scissor rectangle in pixels
+
+    **Returns:** void
+
+    Sets the {{GPURenderPassEncoder/[[scissor_rectangle]]}} used during the rasterization stage.
+    After transformation into viewport coordinates any fragments which fall outside the scissor
+    rectangle will be discarded.
+
+    On the [=Device timeline=], the following steps occur:
```

Comment:
Encoder objects are purely-client-side (content timeline) objects that only serve to encode data that goes into a command buffer object which does exist on the device timeline.

I think `[[scissor_rectangle]]` (and all other state info, like current bind groups, pipeline, vertex buffers, etc.) need to be a local variable in a future "executing a command buffer" algorithm. But for some state info (probably not including scissor rectangle), we need it locally as well to write our validation.

And it *does* make sense to have this text in this location. So I think what we should do is:
- each state-setting command defines a (very short) "state" algorithm that takes an execution state object (in this case, a render pass state object) as a parameter and modifies that state object.
- each non-state-setting command defines an "execution" algorithm that defines what it does when executed based on the current execution state.
- the content-timeline algorithm for a command calls into that command's state algorithm with an encoder-local state object.
- the "executing a command buffer" algorithm calls into both types of algorithms with a state object that's local to "executing a command buffer".

Does this all make sense?
Perhaps we can probably stub this out without too much churn (and without writing "executing a command buffer" or algorithms for most methods yet), but would be OK to do in a separate PR.

Reply:
> Encoder objects are purely-client-side (content timeline) objects that only serve to encode data that goes into a command buffer object which does exist on the device timeline.

Is this actually true for Dawn even? I thought you are sending all the commands to the GPU process, where they get encoded.
In wgpu land, we strongly consider this an implementation detail and want to preserve the right to actually reflect the operations in render pass encoding to the GPU side. Also, the actual validation is done on the wgpu server side today for this, so it's not on a content timeline.

Reply:
Yes, encoder commands are currently device-timeline in Dawn and it would be very good to keep most of the validation on the GPU service side (which can't be content-timeline).

Reply:
Yes, we do it on the service side. You're right it doesn't exactly make sense for this to be content-timeline. But calling it device-timeline isn't really accurate either: we aren't supposed to be able to do things like check bound resource validity here. I guess calling it a device-timeline operation is okay though.

Reply:
Sorry, I guess there is device-timeline state we need for validation, namely the contents of all immutable objects (pipelines, bind groups, etc.) So device-timeline makes sense.

I think the rest of the comment still stands though.

Reply:
Sorry, but I'm not entirely sure what the prescribed changes are here, given that it doesn't seem the timeline needs to change?

Reply:
Sorry, I never figured it out. The main problem is the scissor rectangle is a piece of state that changes over the course of execution of a command buffer, which isn't expressed by a field on the command encoder.

I think really the command encoder needs a list of commands, and calling each of these methods appends a command to that list. Then the scissor rectangle is a piece of local state when executing the list of commands.

However, there is some state that *should* be tracked on the command encoder (during encoding) because it is needed for validation (currently bound bind groups, vertex/index buffers, and pipelines are needed to validate draw/dispatch).

I think what we should do for now is make this more informal and fix it later: Just take `[[scissor_rectangle]]` off the command encoder and use the words "scissor rectangle" informally ("set the scissor rectangle")

Reply:
Done.